### PR TITLE
Refactor chunk handling to separate out stub

### PR DESCRIPTION
**Pull request by cevian**
Previously, the Chunk struct was used to represent both a full
chunk and the stub used for joins. The stub used for joins
only contained valid values for some chunk fields and not others.
After the join determined that a Chunk was complete, it filled
in the rest of the chunk field. The fact that a chunk could have
only some fields filled out and not others at different times,
made the code hard to follow and error prone.

So we separate out the stub state of the chunk into a separate
struct that doesn't contain the not-filled-out fields inside
of it.  This leverages the type system to prevent errors that
try to access invalid fields during the join phase and makes
the code easier to follow.

### CODE REVIEW

File: src/chunk.c [hunk 1]
```diff
@@ -1016,7 +1034,7 @@ chunk_scan_ctx_get_chunk(ChunkScanCtx *ctx)
  * 2) For each dimension slice:
  *	  - Find all chunk constraints matching the dimension slice
  * 3) For each matching chunk constraint
- *	  - Insert a (stub) chunk in a hash table and add the constraint to the chunk
+ *	  - Insert a (chunk) chunk in a hash table and add the constraint to the chunk
```
Comment (JLockerman) [inline src/chunk.c]: huh?

File: src/chunk.c [hunk 2]
```diff
@@ -1210,15 +1230,16 @@ append_chunk_oid(ChunkScanCtx *scanctx, Chunk *chunk)
 }
 
 static ChunkResult
-append_chunk(ChunkScanCtx *scanctx, Chunk *chunk)
+append_chunk(ChunkScanCtx *scanctx, ChunkStub *stub)
```
Comment (JLockerman) [inline src/chunk.c]:
```suggestion
append_chunk_stub(ChunkScanCtx *scanctx, ChunkStub *stub)
```

File: src/chunk.h [hunk 3]
```diff
@@ -46,6 +46,13 @@ typedef struct Chunk
 	ChunkConstraints *constraints;
 } Chunk;
 
+typedef struct ChunkStub
```
Comment (JLockerman) [inline src/chunk.h]: I think this needs a comment explaining what it is conceptually. To my eye it seems like it identifies a subspace of a hyperspace which _could_ be covered by a single chunk.

File: src/chunk.c (hunk 1 above)
Comment (cevian) [reply src/chunk.c]: good catch stupid find/replace

File: src/chunk.c (hunk 2 above)
Comment (cevian) [reply src/chunk.c]: No, this function appends the chunk, not the stub to the data field
